### ISSUE TITLE
Rising/Falling tokens not evenly distributed RE: sample_puzzle 30

### ISSUE THREAD

**Issue by jacobwilliamroy** [chalmerlowe/jarvis_II#17]
So right now, the code for making rising/falling tokens looks something like this:

```
def rising():
    digit = choice( range( 1, 9 ) )
    token = 0

    for i in range( 6 )[::-1]:
        token += digit * (10 ** i)
        digit = choice( range( digit, 10) )

    return token
```
I ran this 1000 times and I noticed the least significant digits of each token tend to group up around 0 if the token is falling or 9 if the token is rising. e.g. 199999 459999 999999 100000 320000 440000 I think this is due to the fact that the range of possible digits is randomly constricted in one direction on each iteration. If there were some way to make a weighted choice in which `choice` is more likely to pick digits closer to the current value of `digit` then I think that would produce a larger variety of rising/falling numbers.

**Comment by chalmerlowe** [comment]
@jacobwilliamroy this is absolutely correct AND it could be construed as a problem. I spent some time contemplating this yesterday while working on the code...

I had two thoughts:

* maybe we don't care, since it doesn't really matter 
* maybe we 'bypass' the problem by randomly grabbing values from the selection of all values

an easy way to get random token values is to simply generate them ALL (000 000 to 999 999) and then randomly select some subset of them for inclusion in the data file.

It would certainly make the generator script much simpler. The 'risk' is that certain patterns will be more common than others and thus if we down selected randomly, we might produce a subset that does not include one OR more categories. Certainly... with a big enough dataset, this won't be problem.

I am pretty tempted to simplify the code... in this way. Do you have any thoughts?

chalmer

**Comment by jacobwilliamroy** [comment]
So I split the number 999999 up into 40 intervals and then ran `map` to measure how often a value from `rising` ends up in each interval if rising is call 10**5 times:
```
[0.02446, 0.02582, 0.01972, 0.03002,
 0.01525, 0.02883, 0.02214, 0.03287,
 0.00466, 0.03283, 0.02434, 0.03637,
 0.00000, 0.02848, 0.02790, 0.04385,
 0.00000, 0.01754, 0.03302, 0.05109,
 0.00000, 0.00000, 0.04030, 0.06006,
 0.00000, 0.00000, 0.02524, 0.07440,
 0.00000, 0.00000, 0.00000, 0.10023,
 0.00000, 0.00000, 0.00000, 0.10086,
 0.00000, 0.00000, 0.00000, 0.10094]
```
You can see how the numbers start off evenly distributed and then drift to the right. I think that drift can be corrected by creating a variable like `possible_values = range( current, 9 )` and then applying some kind of `map` to populate it with extra numbers before choosing an item from it.

As a recreational mathematician, I find this immensely interesting and will happily work on it. However, I do agree that it's not actually necessary to finish preparing the puzzle, and that there are still unfinished things. I could go and work on one of the TODOs you left in the code instead.

**Comment by chalmerlowe** [comment]
As a random thought...
doing a talk at pyhawaii on:
**"How this puzzle series was made: pros, cons, tradeoffs and WTFs"** 

would be a GREAT subject!

**Comment by jacobwilliamroy** [comment]
Here's a very rough guess as to how many there are of each token type:
generic = 10^5
palindromic = 10^3
mini-palindromic = 10^4 
rising/falling <= 8.9 * (10^4)
The only one I'm not at all sure about is the rising/falling tokens. That's really hard to calculate.
